### ISSUE TITLE
Setup HA A-P AZURE

### ISSUE THREAD

**Issue by cbarbudoTF** [fortinetdev/terraform-provider-fortios#91]
I'm unable to setup via terraform an HA Active/passive, to get this CLI in fact:

config system ha
  set group-name "AzureHA"
  set mode a-p
  set hbdev "port3" 100
  set session-pickup enable
  set session-pickup-connectionless enable
  set ha-mgmt-status enable
  config ha-mgmt-interfaces
    edit 1
      set interface "port4"
      set gateway 10.0.4.1
    next
  end
  set override disable
  set priority 255
  set unicast-hb enable
  set unicast-hb-peerip 10.0.3.5
end

Is there any way to get it?

As a workaround I tried to use fortios_system_autoscript calls but are limited to 255 characters so it breaks my config in different blocks.

**Comment by frankshen01** [comment]
Hi @cbarbudoTF   Thank you for reporting this issue, can you please try the configuration as follows? 

```terraform
provider "fortios" {
  hostname = "192.168.52.177"
  token = "GNH7r40H65GNb46kd4rG8rtrmn0fr1"
  insecure = "true"
}

resource "fortios_system_autoscript" "trname1" {
  interval    = 1
  name        = "1"
  output_size = 10
  repeat      = 1
  script      = <<EOF
config system ha
	set group-name "AzureHA"
	set mode a-p
	set hbdev "port3" 100
	set session-pickup enable
	set session-pickup-connectionless enable
	set override disable
	set priority 255
	set unicast-hb enable
	set unicast-hb-peerip 10.0.3.5
end

EOF
  start       = "auto"
}

resource "fortios_system_autoscript" "trname2" {
  interval    = 1
  name        = format("%s%s", fortios_system_autoscript.trname1.name, "extend")
  output_size = 10
  repeat      = 1
  script      = <<EOF
config system ha
	set ha-mgmt-status enable
	config ha-mgmt-interfaces
		edit 1
			set interface "port4"
			set gateway 10.0.4.1
		end
	end
end

EOF
  start       = "auto"
}

```

**validation:**
```
# terraform plan
...
# terraform apply
  ----

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # fortios_system_autoscript.trname1 will be created
  + resource "fortios_system_autoscript" "trname1" {
      + id          = (known after apply)
      + interval    = 1
      + name        = "1"
      + output_size = 10
      + repeat      = 1
      + script      = <<~EOT
            config system ha
                set group-name "AzureHA"
                set mode a-p
                set hbdev "port3" 100
                set session-pickup enable
                set session-pickup-connectionless enable
                set override disable
                set priority 255
                set unicast-hb enable
                set unicast-hb-peerip 10.0.3.5
            end

        EOT
      + start       = "auto"
    }

  # fortios_system_autoscript.trname2 will be created
  + resource "fortios_system_autoscript" "trname2" {
      + id          = (known after apply)
      + interval    = 1
      + name        = "1extend"
      + output_size = 10
      + repeat      = 1
      + script      = <<~EOT
            config system ha
                set ha-mgmt-status enable
                config ha-mgmt-interfaces
                        edit 1
                                set interface "port4"
                                set gateway 10.0.4.1
                        end
                end
            end

        EOT
      + start       = "auto"
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

fortios_system_autoscript.trname1: Creating...
fortios_system_autoscript.trname1: Creation complete after 0s [id=1]
fortios_system_autoscript.trname2: Creating...
fortios_system_autoscript.trname2: Creation complete after 0s [id=1extend]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

myfirewall # config system ha

myfirewall (ha) # show
config system ha
    set group-name "AzureHA"
    set mode a-p
    set hbdev "port3" 100
    set session-pickup enable
    set session-pickup-connectionless enable
    set ha-mgmt-status enable
    config ha-mgmt-interfaces
        edit 1
            set interface "port4"
            set gateway 10.0.4.1
        next
    end
    set override disable
    set priority 255
    set unicast-hb enable
    set unicast-hb-peerip 10.0.3.5
end

myfirewall (ha) #

```
FortiGate does have a limit of 255 for the length of the script. Let‘s’ split it into two parts and keep the order by terraform reference. Please let me know if you have any questions. Thanks!

**Comment by cbarbudoTF** [comment]
Not sure why, but this is what I get:


config system ha
--
set session-pickup enable
set session-pickup-connectionless enable
set session-pickup-expectation enable                  (I added more options...)
set session-pickup-nat enable
set override disable
end

& 


config system auto-script
--
edit "HA"
set interval 1
set start auto
set script "config system ha
set session-pickup enable
set session-pickup-connectionless enable
set session-pickup-expectation enable
set session-pickup-nat enable
set override disable
end
"
next
edit "HAextend"
set interval 1
set start auto
set script "config system probe-response
set http-probe-value OK
set mode http-probe
end
"
next
edit "HAextendextend"
set interval 1
set start auto
set script "config system ha
set group-name AzureHA
set mode a-p
set hbdev port3 100
set ha-mgmt-status enable
set priority 100
set unicast-hb enable
set unicast-hb-peerip 10.160.2.197
end
"
next
edit "HAextendextendextend"
set interval 1
--More--                  set start auto
set script "config system ha
config ha-mgmt-interfaces
edit 1
set interface port1
set gateway 10.160.1.193
next
end
end

(I used 2 more "extend" blocks)
This is my code:

resource "fortios_system_autoscript" "HA" {
  interval    = 1
  name        = "HA"
  output_size = 10
  repeat      = 1
  script      = <<EOF
config system ha
    set session-pickup enable
    set session-pickup-connectionless enable
    set session-pickup-expectation enable
    set session-pickup-nat enable
    set override disable
end
EOF
  start       = "auto"
}

resource "fortios_system_autoscript" "HA2" {
  interval    = 1
  name        = format("%s%s", fortios_system_autoscript.HA.name, "extend")
  output_size = 10
  repeat      = 1
  script      = <<EOF
config system probe-response
    set http-probe-value OK
    set mode http-probe
end
EOF
  start       = "auto"
}

resource "fortios_system_autoscript" "HA3" {
  interval    = 1
  name        = format("%s%s", fortios_system_autoscript.HA2.name, "extend")
  output_size = 10
  repeat      = 1
  script      = <<EOF
config system ha
    set group-name AzureHA
    set mode a-p
    set hbdev port3 100
    set ha-mgmt-status enable
    set priority 100
    set unicast-hb enable
    set unicast-hb-peerip ${var.cluster_sync_peerip}
end
EOF
  start       = "auto"
}

resource "fortios_system_autoscript" "HA4" {
  interval    = 1
  name        = format("%s%s", fortios_system_autoscript.HA3.name, "extend")
  output_size = 10
  repeat      = 1
  script      = <<EOF
config system ha
    config ha-mgmt-interfaces
        edit 1
            set interface port1
            set gateway ${var.on_premises_management_route_gateway}
        next
    end
end
EOF
  start       = "auto"
}

**Comment by frankshen01** [comment]
I ran your script and it worked fine for me. Can you please provide me your FortiGate firemware version? (GUI: Dashboard->Status->System Information->Fireware or CLI: get system status->Version, the version field is enough)

**Comment by cbarbudoTF** [comment]
Fortigate1 # get system status
Version: FortiGate-VM64-AZUREONDEMAND v6.2.4,build1112,200511 (GA)

Thanks in advance!



Carlos Sanchez Barbudo
Consulting & Professional Services Engineer
Spain

T: 
+34 935 241 076
M: 
+34 609217476
E: 
carlos.sanchez@global.ntt
NTT Spain SLU is a subsidiary of NTT Ltd
 NTT Spain SLU with company registration number B62174842 having its registered office at Barcelona and with its registered address at
Calle Pujades, 350 Barcelona 08019.

**Comment by cbarbudoTF** [comment]
And I’m using terraform Forti provider 1.5.3.





Carlos Sanchez Barbudo
Consulting & Professional Services Engineer
Spain

T: 
+34 935 241 076
M: 
+34 609217476
E: 
carlos.sanchez@global.ntt
NTT Spain SLU is a subsidiary of NTT Ltd
 NTT Spain SLU with company registration number B62174842 having its registered office at Barcelona and with its registered address at
Calle Pujades, 350 Barcelona 08019.

**Comment by frankshen01** [comment]
Your configuration does work very well with me. Strange! I re-adjusted your configuration without modifying the parameters. Can you please try the following configuration on your device?

**1 Init status**
```
myFortiGate # get system status
Version: FortiGate-VM64 v6.2.4,build1112,200511 (GA)

myFortiGate # config system ha
myFortiGate (ha) # show
config system ha
    set override disable
end

myFortiGate # config system probe-response
myFortiGate (probe-response) # show
config system probe-response
end

```
**2 Delete old autoscripts**

**3 Terraform config**
```
# cat maintst.tf
provider "fortios" {
  hostname = "192.168.52.177"
  token = "twHgtmQn30gps41388dds8m3drc3rw"
  insecure = "true"
}

variable cluster_sync_peerip {
  type        = string
  default     = "10.0.3.5"
  description = "description"
}

variable on_premises_management_route_gateway {
  type        = string
  default     = "10.0.4.1"
  description = "description"
}

resource "fortios_system_proberesponse" "trname" {
  http_probe_value = "OK"
  mode             = "http-probe"
}

resource "fortios_system_autoscript" "HA1" {
  interval    = 1
  name        = "HA"
  output_size = 10
  repeat      = 1
  script      = <<EOF
config system ha
  set session-pickup enable
  set session-pickup-connectionless enable
  set session-pickup-expectation enable
  set session-pickup-nat enable
  set override disable
  set group-name AzureHA
  set mode a-p
  set hbdev port3 100
end
EOF
  start       = "auto"
}

resource "fortios_system_autoscript" "HA2" {
  interval    = 3
  name        = format("%s%s", fortios_system_autoscript.HA1.name, "ex")
  output_size = 10
  repeat      = 1
  script      = <<EOF
config system ha
  set priority 100
  set unicast-hb enable
  set unicast-hb-peerip ${var.cluster_sync_peerip}
  set ha-mgmt-status enable
  config ha-mgmt-interfaces
    edit 1
      set interface port1
      set gateway ${var.on_premises_management_route_gateway}
    end
  end
end
EOF
  start       = "auto"
}
```

**4 Terraform Apply**
```
# terraform apply
2020/09/29 08:07:08 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # fortios_system_autoscript.HA1 will be created
  + resource "fortios_system_autoscript" "HA1" {
      + id          = (known after apply)
      + interval    = 1
      + name        = "HA"
      + output_size = 10
      + repeat      = 1
      + script      = <<~EOT
            config system ha
              set session-pickup enable
              set session-pickup-connectionless enable
              set session-pickup-expectation enable
              set session-pickup-nat enable
              set override disable
              set group-name AzureHA
              set mode a-p
              set hbdev port3 100
            end
        EOT
      + start       = "auto"
    }

  # fortios_system_autoscript.HA2 will be created
  + resource "fortios_system_autoscript" "HA2" {
      + id          = (known after apply)
      + interval    = 3
      + name        = "HAex"
      + output_size = 10
      + repeat      = 1
      + script      = <<~EOT
            config system ha
              set priority 100
              set unicast-hb enable
              set unicast-hb-peerip 10.0.3.5
              set ha-mgmt-status enable
              config ha-mgmt-interfaces
                edit 1
                  set interface port1
                  set gateway 10.0.4.1
                end
              end
            end
        EOT
      + start       = "auto"
    }

  # fortios_system_proberesponse.trname will be created
  + resource "fortios_system_proberesponse" "trname" {
      + http_probe_value = "OK"
      + id               = (known after apply)
      + mode             = "http-probe"
      + port             = (known after apply)
      + security_mode    = (known after apply)
      + timeout          = (known after apply)
      + ttl_mode         = (known after apply)
    }

Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

fortios_system_proberesponse.trname: Creating...
fortios_system_autoscript.HA1: Creating...
fortios_system_proberesponse.trname: Creation complete after 0s [id=SystemProbeResponse]
fortios_system_autoscript.HA1: Creation complete after 0s [id=HA]
fortios_system_autoscript.HA2: Creating...
fortios_system_autoscript.HA2: Creation complete after 0s [id=HAex]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.

```

**5 Validation**
```
myFortiGate # config system ha

myFortiGate (ha) # show
path=system, objname=ha, tablename=(null), size=5912
config system ha
    set group-name "AzureHA"
    set mode a-p
    set hbdev "port3" 100
    set session-pickup enable
    set session-pickup-connectionless enable
    set session-pickup-expectation enable
    set ha-mgmt-status enable
    config ha-mgmt-interfaces
        edit 1
            set gateway 10.0.4.1
        next
    end
    set override disable
    set priority 100
    set unicast-hb enable
    set unicast-hb-peerip 10.0.3.5
end

myFortiGate # config system probe-response

myFortiGate (probe-response) # show
path=system, objname=probe-response, tablename=(null), size=1166
config system probe-response
    set mode http-probe
end

```
Does this make sense? Thanks!

**Comment by cbarbudoTF** [comment]
Unfortunately, problems with length ¿¿??:

C:\terraform>terraform apply -var-file NEW_IE-primary_LABBK.tfvars

Error: expected length of script to be in the range (0 - 255), got config system ha
  set session-pickup enable
  set session-pickup-connectionless enable
  set session-pickup-expectation enable
  set session-pickup-nat enable
  set override disable
  set group-name AzureHA
  set mode a-p
  set hbdev port3 100
end


  on fortigate_IE.tf line 321, in resource "fortios_system_autoscript" "HA1":
321: resource "fortios_system_autoscript" "HA1" {



Carlos Sanchez Barbudo
Consulting & Professional Services Engineer
Spain

T: 
+34 935 241 076
M: 
+34 609217476
E: 
carlos.sanchez@global.ntt
NTT Spain SLU is a subsidiary of NTT Ltd
 NTT Spain SLU with company registration number B62174842 having its registered office at Barcelona and with its registered address at
Calle Pujades, 350 Barcelona 08019.

**Comment by cbarbudoTF** [comment]
I remove “set override disable” to fit length. It runs but get same problem: Inserted on auto-script !

config system auto-script
    edit "HA"
        set interval 1
        set start auto
        set script "config system ha
  set session-pickup enable
  set session-pickup-connectionless enable
  set session-pickup-expectation enable
  set session-pickup-nat enable
  set group-name AzureHA
  set mode a-p
  set hbdev port3 100
end
"
    next
    edit "HAex"
        set interval 3
--More--                  set start auto
        set script "config system ha
  set priority 100
  set unicast-hb enable
  set unicast-hb-peerip 10.160.2.197
  set ha-mgmt-status enable
  config ha-mgmt-interfaces
    edit 1
      set interface port1
      set gateway 10.160.1.193
    end
  end
end
"


From: Carlos Sanchez Barbudo (Europe)
Sent: martes, 29 de septiembre de 2020 7:29
To: fortinetdev/terraform-provider-fortios <reply@reply.github.com>; fortinetdev/terraform-provider-fortios <terraform-provider-fortios@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: RE: [fortinetdev/terraform-provider-fortios] Setup HA A-P AZURE (#91)

Unfortunately, problems with length ¿¿??:

C:\terraform>terraform apply -var-file NEW_IE-primary_LABBK.tfvars

Error: expected length of script to be in the range (0 - 255), got config system ha
  set session-pickup enable
  set session-pickup-connectionless enable
  set session-pickup-expectation enable
  set session-pickup-nat enable
  set override disable
  set group-name AzureHA
  set mode a-p
  set hbdev port3 100
end


  on fortigate_IE.tf line 321, in resource "fortios_system_autoscript" "HA1":
321: resource "fortios_system_autoscript" "HA1" {



Carlos Sanchez Barbudo
Consulting & Professional Services Engineer
Spain

T: 
+34 935 241 076
M: 
+34 609217476
E: 
carlos.sanchez@global.ntt
NTT Spain SLU is a subsidiary of NTT Ltd
 NTT Spain SLU with company registration number B62174842 having its registered office at Barcelona and with its registered address at
Calle Pujades, 350 Barcelona 08019.

**Comment by cbarbudoTF** [comment]
May be that:

                start       = "auto"

                ,is not working?

                Regards.

From: Carlos Sanchez Barbudo (Europe)
Sent: martes, 29 de septiembre de 2020 7:35
To: fortinetdev/terraform-provider-fortios <reply@reply.github.com>; fortinetdev/terraform-provider-fortios <terraform-provider-fortios@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: RE: [fortinetdev/terraform-provider-fortios] Setup HA A-P AZURE (#91)

                I remove “set override disable” to fit length. It runs but get same problem: Inserted on auto-script !

config system auto-script
    edit "HA"
        set interval 1
        set start auto
        set script "config system ha
  set session-pickup enable
  set session-pickup-connectionless enable
  set session-pickup-expectation enable
  set session-pickup-nat enable
  set group-name AzureHA
  set mode a-p
  set hbdev port3 100
end
"
    next
    edit "HAex"
        set interval 3
--More--                  set start auto
        set script "config system ha
  set priority 100
  set unicast-hb enable
  set unicast-hb-peerip 10.160.2.197
  set ha-mgmt-status enable
  config ha-mgmt-interfaces
    edit 1
      set interface port1
      set gateway 10.160.1.193
    end
  end
end
"


From: Carlos Sanchez Barbudo (Europe)
Sent: martes, 29 de septiembre de 2020 7:29
To: fortinetdev/terraform-provider-fortios <reply@reply.github.com<mailto:reply@reply.github.com>>; fortinetdev/terraform-provider-fortios <terraform-provider-fortios@noreply.github.com<mailto:terraform-provider-fortios@noreply.github.com>>
Cc: Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Subject: RE: [fortinetdev/terraform-provider-fortios] Setup HA A-P AZURE (#91)

Unfortunately, problems with length ¿¿??:

C:\terraform>terraform apply -var-file NEW_IE-primary_LABBK.tfvars

Error: expected length of script to be in the range (0 - 255), got config system ha
  set session-pickup enable
  set session-pickup-connectionless enable
  set session-pickup-expectation enable
  set session-pickup-nat enable
  set override disable
  set group-name AzureHA
  set mode a-p
  set hbdev port3 100
end


  on fortigate_IE.tf line 321, in resource "fortios_system_autoscript" "HA1":
321: resource "fortios_system_autoscript" "HA1" {



Carlos Sanchez Barbudo
Consulting & Professional Services Engineer
Spain

T: 
+34 935 241 076
M: 
+34 609217476
E: 
carlos.sanchez@global.ntt
NTT Spain SLU is a subsidiary of NTT Ltd
 NTT Spain SLU with company registration number B62174842 having its registered office at Barcelona and with its registered address at
Calle Pujades, 350 Barcelona 08019.

**Comment by frankshen01** [comment]
Hi, @cbarbudoTF , Sorry I misunderstood what you mean earlier. What you care about is why there are several items in auto-script. Rather than whether HA is configured successfully. I wonder if my understanding is correct now? Just want to confirm with you: has the HA configuration succeeded through the new script?

1. Multiple scripts inserted is expected, since FortiGate relies on these scripts item to run once immediately to configure HA after these scripts are created.  These scripts are only run once after they are created and they will remain there until they are actively deleted, and terraform destroy can delete them also, You can choose to delete them or keep them there.

2. The reason for the problem of  length is that the **_Tab_** before each line in the script is converted into _**multiple spaces**_ after being copied to github, so the length is too long when you paste it again. Just delete the preceding spaces or change multiple spaces to a single tab as following, it will be ok:

```
resource "fortios_system_autoscript" "HA1" {
  interval = 1
  name = "HA"
  output_size = 10
  repeat = 1
  script = <<EOF
config system ha
set session-pickup enable
set session-pickup-connectionless enable
set session-pickup-expectation enable
set session-pickup-nat enable
set override disable
set group-name AzureHA
set mode a-p
set hbdev port3 100
end
EOF
  start = "auto"
}

resource "fortios_system_autoscript" "HA2" {
  interval = 3
  name = format("%s%s", fortios_system_autoscript.HA1.name, "ex")
  output_size = 10
  repeat = 1
  script = <<EOF
config system ha
set priority 100
set unicast-hb enable
set unicast-hb-peerip ${var.cluster_sync_peerip}
set ha-mgmt-status enable
config ha-mgmt-interfaces
edit 1
set interface port1
set gateway ${var.on_premises_management_route_gateway}
end
end
end
EOF
  start = "auto"
}

```
BTW: I deleted the footnote of your email to prevent your information from being leaked. Thanks!

**Comment by cbarbudoTF** [comment]
(Thanks about deleting private info).

No, it failed. This is only cli I get:


config system ha
 set override disable
 set priority 100
end

I checked scripts where executed:


Fortigate1 # execute auto-script status
--
HA <-- executed, output file size: 722
HAex <-- executed, output file size: 650

**Comment by frankshen01** [comment]
I wonder if you can contact the Fortigate SE who serves you to provide on-site or online services? Because I successfully ran the above script on three devices, the problem could not be reproduced. I will work with them to solve this problem.

**Comment by cbarbudoTF** [comment]
I'll try it.
FYI, i'm checking that i.e. if my script include

set session-pickup enable
set session-pickup-connectionless enable
set session-pickup-expectation enable
set session-pickup-nat enable
set override disable

it works.

if I add..

set group-name AzureHA
set mode a-p

It doesn't. I'll continue testing.

Regards.

**Comment by frankshen01** [comment]
BTW, on your current device, can you successfully configure HA with these parameters directly from the CLI?

**Comment by cbarbudoTF** [comment]
Yes. At least I tried "set mode a-p" & no problem. 
I'm repeating deployment adding just one command each time to get more info.
At this moment, I got:

config system ha
--
set group-name "AzureHA"
set hbdev "port3" 100
set session-pickup enable
set session-pickup-connectionless enable
set session-pickup-expectation enable
set session-pickup-nat enable
set override disable
set priority 100
end

Problems with "set mode a-p" & "set unicast-hb enable". I'm still testing.
